sdk/eventhub/event-hubs/samples/typescript/src/iothubConnectionString.ts

```diff
@@ -0,0 +1,166 @@
+): string {
+  resourceUri = encodeURIComponent(resourceUri);
+
+  // Set expiration in seconds.
+  const expires = Math.ceil(Date.now() / 1000 + expiresInMins * 60);
+  const toSign = resourceUri + "\n" + expires;
+
+  // Use the crypto module to create the hmac.
+  const hmac = crypto.createHmac("sha256", Buffer.from(signingKey, "base64"));
+  hmac.update(toSign);
+  const base64UriEncoded = encodeURIComponent(hmac.digest("base64"));
+
+  // Construct authorization string.
+  return `SharedAccessSignature sr=${resourceUri}&sig=${base64UriEncoded}&se=${expires}&skn=${policyName}`;
+}
+
+/**
+ * Converts an IotHub Connection string into an Event Hubs-compatible connection string.
+ * @param connectionString An IotHub connection string in the format:
+ * `"HostName=<your-iot-hub>.azure-devices.net;SharedAccessKeyName=<KeyName>;SharedAccessKey=<Key>"`
+ * @returns An Event Hubs-compatible connection string in the format:
+ * `"Endpoint=sb://<hostname>;EntityPath=<your-iot-hub>;SharedAccessKeyName=<KeyName>;SharedAccessKey=<Key>"`
+ */
+async function convertIotHubToEventHubsConnectionString(connectionString: string): Promise<string> {
+  const { HostName, SharedAccessKeyName, SharedAccessKey } = parseConnectionString<{
+    HostName: string;
+    SharedAccessKeyName: string;
+    SharedAccessKey: string;
+  }>(connectionString);
+
+  // Verify that the required info is in the connection string.
+  if (!HostName || !SharedAccessKey || !SharedAccessKeyName) {
+    throw new Error(`Invalid IotHub connection string.`);
+  }
+
+  //Extract the IotHub name from the hostname.
+  const [iotHubName] = HostName.split(".");
+
+  if (!iotHubName) {
+    throw new Error(`Unable to extract the IotHub name from the connection string.`);
+  }
+
+  // Generate a token to authenticate to the service.
+  // The code for generateSasToken can be found at https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-tokens
+  const token = generateSasToken(
+    `${HostName}/messages/events`,
+    SharedAccessKey,
+    SharedAccessKeyName,
+    5 // token expires in 5 minutes
+  );
+
+  const connection = new Connection({
+    transport: "tls",
+    host: HostName,
+    hostname: HostName,
+    username: `${SharedAccessKeyName}@sas.root.${iotHubName}`,
+    port: 5671,
+    reconnect: false,
+    password: token
+  });
+  await connection.open();
+
+  // Create the receiver that will trigger a redirect error.
+  const receiver = await connection.createReceiver({
+    source: { address: `amqps://${HostName}/messages/events/$management` }
+  });
+
+  return new Promise((resolve, reject) => {
+    receiver.on(ReceiverEvents.receiverError, (context) => {
+      const error = context.receiver && context.receiver.error;
+      if (isAmqpError(error) && error.condition === "amqp:link:redirect") {
+        const hostname = error.info && error.info.hostname;
+        if (!hostname) {
+          reject(new Error(`Unable to convert connection string.`));
```

Comment:
Is that what's happening here? Also, should you forward on any information from `error` or is it useless for this particular case?

Reply:
I can just throw the original error in this case.